File: sdk/swift/Sources/FluxChat/KnowledgeClient.swift
import Foundation

/// Helper HTTP interne utilisé par `FluxChatClient` et `KnowledgeClient`.
actor HTTPHelper {
    let apiKey: String
    let baseURL: String
    let session: URLSession

    init(apiKey: String, baseURL: String, session: URLSession = .shared) {
        self.apiKey = apiKey
        self.baseURL = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        self.session = session
    }

    // MARK: - Public HTTP methods

    func get<T: Decodable>(_ path: String, jwtToken: String? = nil) async throws -> T {
        let req = try buildRequest(method: "GET", path: path, body: nil as String?, jwtToken: jwtToken)
        return try await performEnveloped(req)
    }

    func post<B: Encodable, T: Decodable>(_ path: String, body: B, jwtToken: String? = nil) async throws -> T {
        let req = try buildRequest(method: "POST", path: path, body: body, jwtToken: jwtToken)
        return try await performEnveloped(req)
    }

    func postVoid<B: Encodable>(_ path: String, body: B, jwtToken: String? = nil) async throws {
        let req = try buildRequest(method: "POST", path: path, body: body, jwtToken: jwtToken)
        let (_, response) = try await fetch(req)
        try validateStatus(response, data: Data())
    }

    func patch<B: Encodable, T: Decodable>(_ path: String, body: B, jwtToken: String? = nil) async throws -> T {
        let req = try buildRequest(method: "PATCH", path: path, body: body, jwtToken: jwtToken)
        return try await performEnveloped(req)
    }

    func delete(_ path: String, jwtToken: String? = nil) async throws {
        let req = try buildRequest(method: "DELETE", path: path, body: nil as String?, jwtToken: jwtToken)
        let (_, response) = try await fetch(req)
        try validateStatus(response, data: Data())
    }

    // MARK: - Private helpers

    private func buildRequest<B: Encodable>(
        method: String, path: String, body: B?, jwtToken: String?
    ) throws -> URLRequest {
        guard let url = URL(string: baseURL + path) else {
            throw FluxChatNetworkError("Invalid URL: \(baseURL + path)")
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")

        // Auth routing: JWT for admin routes, X-API-Key for public routes
        if let jwt = jwtToken {
            req.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")
        } else {
            req.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        }

        if let body {
            do { req.httpBody = try JSONEncoder().encode(body) }
            catch { throw FluxChatNetworkError("Failed to encode request body", underlying: error) }
        }
        return req
    }

    /// Decodes a response wrapped in { "success": true, "data": ... }
    private func performEnveloped<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, response) = try await fetch(req)
        try validateStatus(response, data: data)
        let decoder = JSONDecoder()
        do {
            let envelope = try decoder.decode(APIEnvelope<T>.self, from: data)
            if let payload = envelope.data {
                return payload
            }
            throw FluxChatNetworkError("Response envelope contained no data")
        } catch let e as FluxChatNetworkError { throw e }
        catch { throw FluxChatNetworkError("Failed to decode response", underlying: error) }
    }

    private func fetch(_ req: URLRequest) async throws -> (Data, HTTPURLResponse) {
        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw FluxChatNetworkError("Non-HTTP response received")
            }
            return (data, http)
        } catch let e as FluxChatNetworkError { throw e }
        catch { throw FluxChatNetworkError("Network request failed", underlying: error) }
    }

    private func validateStatus(_ response: HTTPURLResponse, data: Data) throws {
        let code = response.statusCode
        guard (200..<300).contains(code) else {
            // Try to extract error message from envelope
            let apiMsg: String
            if let env = try? JSONDecoder().decode(APIEnvelope<String>.self, from: data),
               let msg = env.message {
                apiMsg = msg
            } else {
                apiMsg = HTTPURLResponse.localizedString(forStatusCode: code)
            }
            throw FluxChatApiError(statusCode: code, apiMessage: apiMsg)
        }
    }
}

/// Client Knowledge CRUD, accessible via `FluxChatClient.knowledge`.
public final class KnowledgeClient {
    let http: HTTPHelper
    private let jwtToken: String?

    init(http: HTTPHelper, jwtToken: String? = nil) {
        self.http = http
        self.jwtToken = jwtToken
    }

    /// Liste tous les éléments de la base de connaissance (requiert un JWT).
    public func list() async throws -> [KnowledgeItem] {
        try await http.get("/bot/knowledge", jwtToken: jwtToken)
    }

    /// Récupère un élément par son ID (requiert un JWT).
    public func get(id: String) async throws -> KnowledgeItem {
        try await http.get("/bot/knowledge/\(id)", jwtToken: jwtToken)
    }

    /// Crée un nouvel élément (requiert un JWT).
    public func create(
        title: String,
        content: String,
        category: String? = nil,
        keywords: [String]? = nil
    ) async throws -> KnowledgeItem {
        let body = KnowledgeCreateRequest(
            title: title,
            content: content,
            category: category,
            keywords: keywords
        )
        return try await http.post("/bot/knowledge", body: body, jwtToken: jwtToken)
    }

    /// Met à jour un élément existant avec un patch partiel (requiert un JWT).
    public func update(
        id: String,
        title: String? = nil,
        content: String? = nil,
        category: String? = nil,
        keywords: [String]? = nil,
        isActive: Bool? = nil
    ) async throws -> KnowledgeItem {
        let body = KnowledgePatchRequest(
            title: title,
            content: content,
            category: category,
            keywords: keywords,
            isActive: isActive
        )
        return try await http.patch("/bot/knowledge/\(id)", body: body, jwtToken: jwtToken)
    }

    /// Supprime un élément par son ID (requiert un JWT).
    public func delete(id: String) async throws {
        try await http.delete("/bot/knowledge/\(id)", jwtToken: jwtToken)
    }
}

File: sdk/dart/lib/src/knowledge_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'exceptions.dart';
import 'models.dart';

/// Client fluent pour les opérations Knowledge Base.
/// Accédez-y via [FluxChat.knowledge].
class KnowledgeClient {
  final _HttpHelper _http;
  final String jwtToken;

  KnowledgeClient(this._http, {required this.jwtToken});

  /// Retourne tous les éléments de la base de connaissance.
  Future<List<KnowledgeItem>> list() async {
    final data = await _http.get('/bot/knowledge', jwtToken: jwtToken) as List<dynamic>;
    return data
        .map((e) => KnowledgeItem.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Retourne un élément par son identifiant.
  Future<KnowledgeItem> get(String id) async {
    final data = await _http.get('/bot/knowledge/$id', jwtToken: jwtToken) as Map<String, dynamic>;
    return KnowledgeItem.fromJson(data);
  }

  /// Crée un nouvel élément de connaissance.
  Future<KnowledgeItem> create(
    String title,
    String content, {
    String? category,
    List<String>? keywords,
  }) async {
    final payload = <String, dynamic>{
      'title': title,
      'content': content,
    };
    if (category != null) payload['category'] = category;
    if (keywords != null) payload['keywords'] = keywords;

    final data = await _http.post('/bot/knowledge', payload, jwtToken: jwtToken) as Map<String, dynamic>;
    return KnowledgeItem.fromJson(data);
  }

  /// Met à jour un élément existant (mise à jour partielle).
  Future<KnowledgeItem> update(
    String id, {
    String? title,
    String? content,
    String? category,
    List<String>? keywords,
    bool? isActive,
  }) async {
    final payload = <String, dynamic>{};
    if (title != null) payload['title'] = title;
    if (content != null) payload['content'] = content;
    if (category != null) payload['category'] = category;
    if (keywords != null) payload['keywords'] = keywords;
    if (isActive != null) payload['isActive'] = isActive;

    final data = await _http.patch('/bot/knowledge/$id', payload, jwtToken: jwtToken) as Map<String, dynamic>;
    return KnowledgeItem.fromJson(data);
  }

  /// Supprime un élément par son identifiant.
  Future<void> delete(String id) async {
    await _http.delete('/bot/knowledge/$id', jwtToken: jwtToken);
  }
}

/// Helper HTTP interne partagé entre [FluxChat] et [KnowledgeClient].
class _HttpHelper {
  final String apiKey;
  final String baseUrl;
  final http.Client httpClient;

  _HttpHelper({
    required this.apiKey,
    required this.baseUrl,
    required this.httpClient,
  });

  Map<String, String> _headers(String? jwtToken) {
    final h = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
    if (jwtToken != null) {
      h['Authorization'] = 'Bearer $jwtToken';
    } else {
      h['X-API-Key'] = apiKey;
    }
    return h;
  }

  Future<dynamic> get(String path, {String? jwtToken}) => _send('GET', path, jwtToken: jwtToken);
  
  Future<dynamic> post(String path, Map<String, dynamic> body, {String? jwtToken}) =>
      _send('POST', path, body: body, jwtToken: jwtToken);
      
  Future<void> postVoid(String path, Map<String, dynamic> body, {String? jwtToken}) async {
      await _send('POST', path, body: body, jwtToken: jwtToken, expectData: false);
  }

  Future<dynamic> patch(String path, Map<String, dynamic> body, {String? jwtToken}) =>
      _send('PATCH', path, body: body, jwtToken: jwtToken);

  Future<void> delete(String path, {String? jwtToken}) async => 
      _send('DELETE', path, jwtToken: jwtToken, expectData: false);

  Future<dynamic> _send(String method, String path,
      {Map<String, dynamic>? body, String? jwtToken, bool expectData = true}) async {
    final uri = Uri.parse('$baseUrl$path');
    http.Response response;
    final headers = _headers(jwtToken);

    try {
      switch (method) {
        case 'GET':
          response = await httpClient.get(uri, headers: headers);
          break;
        case 'POST':
          response = await httpClient.post(uri,
              headers: headers, body: jsonEncode(body));
          break;
        case 'PATCH':
          response = await httpClient.patch(uri,
              headers: headers, body: jsonEncode(body));
          break;
        case 'DELETE':
          response = await httpClient.delete(uri, headers: headers);
          break;
        default:
          throw UnsupportedError('HTTP method $method not supported');
      }
    } catch (e) {
      if (e is FluxChatApiException || e is FluxChatNetworkException) rethrow;
      throw FluxChatNetworkException('Request failed for $method $path', e);
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      String msg = response.body;
      try {
          final decoded = jsonDecode(response.body) as Map<String, dynamic>;
          if (decoded.containsKey('message') && decoded['message'] != null) {
              msg = decoded['message'].toString();
          }
      } catch (_) {}
      throw FluxChatApiException(response.statusCode, msg);
    }

    if (response.body.isEmpty || response.body == 'null') return null;
    
    final decoded = jsonDecode(response.body);
    if (decoded is Map<String, dynamic> && decoded.containsKey('success')) {
        if (expectData) {
           return decoded['data'];
        }
        return null;
    }
    
    return decoded;
  }
}
